src/dimension/Dimension.js

```diff
@@ -0,0 +1,488 @@
+                subExpression += dimension + " >= " + typedRange[0] + " AND " + dimension + " <= " + typedRange[1]
+            } else {
+                subExpression += _dimArray[e] + " >= " + typedRange[0] + " AND " + _dimArray[e] + " <= " + typedRange[1]
+            }
+        }
+        if (inverseFilters) {
+            subExpression = "NOT(" + subExpression + ")"
+        }
+        if (append) {
+            this._scopedFilters[_dimensionIndex] += "(" + subExpression + ")"
+        } else {
+            this._scopedFilters[_dimensionIndex] = "(" + subExpression + ")"
+        }
+        return this
+    }
+    formatRelativeValue(val) {
+        if (val.now) {
+            return "NOW()" // todo - there might be subtle bugs in Now(), depending on when it's called & evaluated
+        } else if (val.datepart && typeof val.number !== 'undefined') {
+            const date      = typeof val.date !== 'undefined' ? val.date : "NOW()",
+                operator    = typeof val.operator !== 'undefined' ? val.operator : "DATE_ADD",
+                number      = isNaN(val.number) ? this.formatRelativeValue(val.number) : val.number,
+                add         = typeof val.add !== 'undefined' ? val.add : ""
+            return `${operator}(${val.datepart}, ${number}, ${date})${add}`
+        } else {
+            return val
+        }
+    }
+    filterMulti(filterArray, resetRangeIn, inverseFilters, binParams) {
+        const { _dimensionIndex, _drillDownFilter } = this,
+            crossfilter                             = this.getCrossfilter()
+        let resetRange = false
+        if (resetRangeIn !== undefined) {
+            resetRange = resetRangeIn
+        }
+        const lastFilterIndex = filterArray.length - 1
+        this._scopedFilters[_dimensionIndex] = "("
+
+        inverseFilters = typeof (inverseFilters) === "undefined" ? false : inverseFilters
+
+        filterArray.forEach((currentFilter, i) => {
+            this.filter(currentFilter, true, resetRange, inverseFilters, binParams)
+            if (i !== lastFilterIndex) {
+                if (_drillDownFilter ^ inverseFilters) {
+                    crossfilter._filters[_dimensionIndex] += " AND "
+                } else {
+                    crossfilter._filters[_dimensionIndex] += " OR "
+                }
+            }
+        })
+        this._scopedFilters[_dimensionIndex] += ")"
+        return this
+    }
+    filterAll(softFilterClear) {
+        const { _dimensionIndex } = this
+        if (softFilterClear === undefined || softFilterClear === false) {
+            this._rangeFilters = []
+        }
+        this._filterVal = null
+        this._scopedFilters[_dimensionIndex] = ""
+        return this
+    }
+    samplingRatio(ratio) {
+        if (!ratio)
+            this._samplingRatio = null
+        this._samplingRatio = ratio // TODO always overwrites; typo?
```

Comment:
I think this is a typo. Does it break things if it's fixed? also this looks like our usual setter/getter pattern, but it doesn't behave the way the others do. I think it should be consistent with the other patterns, ie invoked with zero args returns value, invoked with arg sets value to arg and returns this.